fuel-relayer/src/finalization_queue.rs

```diff
@@ -0,0 +1,521 @@
+    blocks: PendingBlocks,
+    /// Current validator set
+    validators: Validators,
+}
+
+/// Pending diff between FuelBlocks
+#[derive(Clone, Debug, Default)]
+pub struct DaBlockDiff {
+    /// da block height
+    pub da_height: DaBlockHeight,
+    /// Validator stake deposit and withdrawel.
+    pub validators: HashMap<Address, Option<Address>>,
+    // Delegation diff contains new delegation list, if we did just withdrawal option will be None.
+    pub delegations: HashMap<Address, Option<HashMap<Address, u64>>>,
+    /// erc-20 pending deposit.
+    pub assets: HashMap<Bytes32, AssetDepositLog>,
+}
+
+impl DaBlockDiff {
+    pub fn new(da_height: DaBlockHeight) -> Self {
+        Self {
+            da_height,
+            validators: HashMap::new(),
+            delegations: HashMap::new(),
+            assets: HashMap::new(),
+        }
+    }
+}
+
+impl FinalizationQueue {
+    pub fn new(
+        chain_id: u64,
+        contract_address: H160,
+        private_key: &[u8],
+        last_commited_finalized_fuel_height: BlockHeight,
+    ) -> Self {
+        let blocks = PendingBlocks::new(
+            chain_id,
+            contract_address,
+            private_key,
+            last_commited_finalized_fuel_height,
+        );
+        Self {
+            blocks,
+            pending: VecDeque::new(),
+            validators: Validators::default(),
+            bundled_removed_eth_events: Vec::new(),
+            finalized_da_height: 0,
+        }
+    }
+
+    pub async fn load_validators(&mut self, db: &dyn RelayerDb) {
+        self.validators.load(db).await
+    }
+
+    pub async fn get_validators(
+        &mut self,
+        da_height: DaBlockHeight,
+    ) -> Option<HashMap<Address, (u64, Option<Address>)>> {
+        self.validators.get(da_height).await
+    }
+
+    pub fn clear(&mut self) {
+        self.pending.clear()
+    }
+
+    /// Bundle all removed events to apply them in same time when all of them are flushed.
+    fn bundle_removed_events(&mut self, event: EthEventLog, da_height: DaBlockHeight) {
+        // agregate all removed events before reverting them.
+        // check if we have pending block for removal
+        if let Some((last_eth_block, list)) = self.bundled_removed_eth_events.last_mut() {
+            // check if last pending block is same as log event that we received.
+            if *last_eth_block == da_height {
+                list.push(event)
+            } else {
+                // if block number differs just push new block.
+                self.bundled_removed_eth_events
+                    .push((da_height, vec![event]));
+            }
+        } else {
+            // if there are not pending block for removal just add it.
+            self.bundled_removed_eth_events
+                .push((da_height, vec![event]));
+        }
+    }
+
+    /// propagate new fuel block to pending_blocks
+    pub fn handle_fuel_block(&mut self, block: &Arc<SealedFuelBlock>) {
```

Comment:
`&Arc` is an anti-pattern imo. Since Arc impls `Borrow<T>`, if the block comes from an arc we can just borrow it before passing to this function.

```suggestion
    pub fn handle_fuel_block(&mut self, block: &SealedFuelBlock) {
```